### PR TITLE
network: use ordered_set_ensure_put

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/network/networkd-route.c [hunk 1]
```diff
@@ -2341,12 +2341,10 @@ int config_parse_multipath_route(
...
 
-        r = ordered_set_ensure_allocated(&n->multipath_routes, NULL);
-        if (r < 0)
+        r = ordered_set_ensure_put(&n->multipath_routes, NULL, m);
+        if (r == -ENOMEM)
                 return log_oom();
-
-        r = ordered_set_put(n->multipath_routes, m);
-        if (r < 0) {
+        else if (r < 0) {
```
Comment (poettering) [inline src/network/networkd-route.c]: doesn't matter, but you could replace "else if" by just "if" here to keep things short, given that the first branch returns from the function

File: src/network/networkd-ndisc.c [hunk 2]
```diff
@@ -1544,12 +1544,10 @@ int config_parse_address_generation_type(
...
 
-        r = ordered_set_ensure_allocated(&network->ipv6_tokens, &ipv6_token_hash_ops);
-        if (r < 0)
+        r = ordered_set_ensure_put(&network->ipv6_tokens, &ipv6_token_hash_ops, token);
+        if (r == -ENOMEM)
                 return log_oom();
-
-        r = ordered_set_put(network->ipv6_tokens, token);
-        if (r == -EEXIST)
+        else if (r == -EEXIST)
```
Comment (poettering) [inline src/network/networkd-ndisc.c]: ditto

**Comment by ssahani** [comment]
updated. thanks for the review . ptal

**Comment by bluca** [comment]
There's a memory leak reported from the same function that was changede:

```
Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x4a517d in malloc /src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:145:3
    #1 0x5dcad0 in malloc_multiply /work/build/../../src/systemd/src/basic/alloc-util.h:96:16
    #2 0x5dca4e in ipv6token_new /work/build/../../src/systemd/src/network/networkd-ndisc.c:1379:13
    #3 0x5dd633 in config_parse_address_generation_type /work/build/../../src/systemd/src/network/networkd-ndisc.c:1506:13
    #4 0x7fce34e0ffbe in next_assignment /work/build/../../src/systemd/src/shared/conf-parser.c:135:32
    #5 0x7fce34e090f9 in parse_line /work/build/../../src/systemd/src/shared/conf-parser.c:242:16
    #6 0x7fce34e0848d in config_parse /work/build/../../src/systemd/src/shared/conf-parser.c:378:21
    #7 0x7fce34e09a88 in config_parse_many_files /work/build/../../src/systemd/src/shared/conf-parser.c:440:21
    #8 0x7fce34e0a1a8 in config_parse_many /work/build/../../src/systemd/src/shared/conf-parser.c:508:16
    #9 0x4ecf44 in network_load_one /work/build/../../src/systemd/src/network/networkd-network.c:426:13
    #10 0x4d80a4 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/network/fuzz-network-parser.c:26:16
    #11 0x6d35a6 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:599:15
    #12 0x6d1551 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/libfuzzer/FuzzerLoop.cpp:505:3
    #13 0x6d6451 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/libfuzzer/FuzzerLoop.cpp:814:7
    #14 0x6d6dd5 in fuzzer::Fuzzer::Loop(std::__1::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/libfuzzer/FuzzerLoop.cpp:845:3
    #15 0x695c57 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:906:6
    #16 0x6871c7 in main /src/libfuzzer/FuzzerMain.cpp:20:10
    #17 0x7fce33b5383f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)

DEDUP_TOKEN: malloc--malloc_multiply--ipv6token_new
SUMMARY: AddressSanitizer: 24 byte(s) leaked in 1 allocation(s).
```

**Comment by ssahani** [comment]
dropped the commit for now. will follow up with another set

**Comment by bluca** [comment]
Could this test failure be related? I haven't seen it before:

```
484/544 test-dhcp-server                        TIMEOUT 30.37 s 

--- command ---
SYSTEMD_KBD_MODEL_MAP='/tmp/autopkgtest.SzT5aW/build.BgU/systemd/src/locale/kbd-model-map' PATH='/tmp/autopkgtest.SzT5aW/build.BgU/systemd/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin' SYSTEMD_LANGUAGE_FALLBACK_MAP='/tmp/autopkgtest.SzT5aW/build.BgU/systemd/src/locale/language-fallback-map' /tmp/autopkgtest.SzT5aW/build.BgU/systemd/build-deb/test-dhcp-server
--- stderr ---
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
Assertion '!server->event' failed at src/libsystemd-network/sd-dhcp-server.c:198, function sd_dhcp_server_attach_event(). Ignoring.
Assertion '!server->event' failed at src/libsystemd-network/sd-dhcp-server.c:198, function sd_dhcp_server_attach_event(). Ignoring.
Assertion 'server->address != htobe32(INADDR_ANY)' failed at src/libsystemd-network/sd-dhcp-server.c:1021, function sd_dhcp_server_start(). Ignoring.
Assertion 'address->s_addr != INADDR_ANY' failed at src/libsystemd-network/sd-dhcp-server.c:50, function sd_dhcp_server_configure_pool(). Ignoring.
Assertion 'prefixlen <= 32' failed at src/libsystemd-network/sd-dhcp-server.c:51, function sd_dhcp_server_configure_pool(). Ignoring.

```

**Comment by ssahani** [comment]
> Could this test failure be related? I haven't seen it before:
> 
> ```
> 484/544 test-dhcp-server                        TIMEOUT 30.37 s 
> 
> --- command ---
> SYSTEMD_KBD_MODEL_MAP='/tmp/autopkgtest.SzT5aW/build.BgU/systemd/src/locale/kbd-model-map' PATH='/tmp/autopkgtest.SzT5aW/build.BgU/systemd/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin' SYSTEMD_LANGUAGE_FALLBACK_MAP='/tmp/autopkgtest.SzT5aW/build.BgU/systemd/src/locale/language-fallback-map' /tmp/autopkgtest.SzT5aW/build.BgU/systemd/build-deb/test-dhcp-server
> --- stderr ---
> Failed to read $container of PID 1, ignoring: Permission denied
> Found container virtualization none.
> Assertion '!server->event' failed at src/libsystemd-network/sd-dhcp-server.c:198, function sd_dhcp_server_attach_event(). Ignoring.
> Assertion '!server->event' failed at src/libsystemd-network/sd-dhcp-server.c:198, function sd_dhcp_server_attach_event(). Ignoring.
> Assertion 'server->address != htobe32(INADDR_ANY)' failed at src/libsystemd-network/sd-dhcp-server.c:1021, function sd_dhcp_server_start(). Ignoring.
> Assertion 'address->s_addr != INADDR_ANY' failed at src/libsystemd-network/sd-dhcp-server.c:50, function sd_dhcp_server_configure_pool(). Ignoring.
> Assertion 'prefixlen <= 32' failed at src/libsystemd-network/sd-dhcp-server.c:51, function sd_dhcp_server_configure_pool(). Ignoring.
> ```

it's ignoring .

**Review by bluca — APPROVED** [review]
lgtm

**Comment by bluca** [comment]
> lgtm

But 3 autopkgtest failure in a row is quite bad luck - could you please do a rebase and push so that it runs again? Just for peace of mind

**Pull request merged**